module/module.go

```diff
@@ -0,0 +1,501 @@
+						filepath.Join(class.Directory, file.Name()),
+					)
+					if instanceErr != nil {
+						return nil, errors.Wrapf(
+							instanceErr,
+							"Error reading multi instance %s in %s",
+							className,
+							filepath.Join(class.Directory, file.Name()),
+						)
+					}
+					classInstances = append(classInstances, instance)
+				}
+			}
+		}
+
+		resolvedModules[className] = classInstances
+	}
+
+	return resolvedModules, nil
+}
+
+func readInstance(
+	className string,
+	baseDirectory string,
+	instanceDirectory string,
+) (*Instance, error) {
+
+	jsonFileName := className + configSuffix
+	classConfigPath := filepath.Join(
+		baseDirectory,
+		instanceDirectory,
+		jsonFileName,
+	)
+
+	jsonConfig := JSONClassConfig{}
+	raw, err := jsonConfig.Read(classConfigPath)
+
+	if err != nil {
+		// TODO: We should accumulate errors and list them all here
+		// Expected $class-config.json to exist in ...
+		return nil, errors.Wrapf(
+			err,
+			"Error reading JSON Config %s",
+			classConfigPath,
+		)
+	}
+
+	return &Instance{
+		ClassName:     className,
+		ClassType:     jsonConfig.Type,
+		BaseDirectory: baseDirectory,
+		Directory:     instanceDirectory,
+		InstanceName:  jsonConfig.Name,
+		Dependencies:  readDeps(jsonConfig.Dependencies),
+		JSONFileName:  jsonFileName,
+		JSONFileRaw:   raw,
+	}, nil
+}
+
+func readDeps(jsonDeps map[string][]string) []Dependency {
+	depCount := 0
+
+	for _, depsList := range jsonDeps {
+		depCount += len(depsList)
+	}
+
+	deps := make([]Dependency, depCount)
+	depIndex := 0
+
+	for className, depsList := range jsonDeps {
+		for _, instanceName := range depsList {
+			deps[depIndex] = Dependency{
+				ClassName:    className,
+				InstanceName: instanceName,
+			}
+			depIndex++
+		}
+	}
+
+	return deps
+}
+
+// GenerateBuild will, given a module system configuration directory and a
+// target build directory, run the generators assigned to each type of module
+// and write the generated output to the module build directory
+func (moduleSystem *System) GenerateBuild(baseDirectory string) error {
+	resolvedModules, err := moduleSystem.ResolveModules(baseDirectory)
+
+	if err != nil {
+		return err
+	}
+
+	moduleCount := 0
+	for _, moduleList := range resolvedModules {
+		moduleCount += len(moduleList)
+	}
+
+	moduleIndex := 0
+	for _, className := range moduleSystem.classOrder {
+		classInstances := resolvedModules[className]
+
+		for _, classInstance := range classInstances {
+			moduleIndex++
+			buildPath := filepath.Join(
+				classInstance.BaseDirectory,
+				"build",
+				classInstance.Directory,
+			)
+			fmt.Printf(
+				"Generating %s %s %s in %s %d/%d\n",
+				classInstance.ClassType,
+				classInstance.ClassName,
+				classInstance.InstanceName,
+				buildPath,
+				moduleIndex,
+				moduleCount,
+			)
+
+			classGenerators := moduleSystem.classes[classInstance.ClassName]
+			generator := classGenerators.types[classInstance.ClassType]
+
+			if generator == nil {
+				fmt.Printf(
```

Comment:
It's a question of whether or not the module system defined has to be exhaustive or not. In this case I am skipping "custom" clients. It originally caused a panic after rebasing on a change that added that client type. I think it makes sense to assume module directories are "fully managed" and therefore this would be an error in future.